[AArch64] Make IFUNC opt-in rather than opt-out.

hvdijk opened the pull request:
IFUNCs require loader support, so for arbitrary environments, the safe assumption is to assume that they are not supported. In particular, aarch64-linux-pauthtest may be used with musl, and was wrongly detected as supporting IFUNCs.

With IFUNC support now being detected more reliably, this also removes the check for PAuth support. If both are supported, either would work.

hvdijk commented:
cc @pcc, this changes #133533, does this look okay to you? And if so, do you think it needs a test?

hvdijk commented:
Oh, I see a problem. From `targetSupportsPAuthRelocation`: "No released version of glibc supports PAuth relocations."

So basically, we need to disable PAuth relocations for glibc, but we also need to disable IFUNC for anything not glibc. What target is meant to be handled here?

Either way, even without this PR, #133533 looks like it has a problem, it results in a crash if the existing test is updated to use `aarch64-linux-gnu`.

pcc commented:
> I do not think musl is alone in not supporting it, I see no support for it in uclibc-ng either.

Yeah, there are some others that don't support it.

> What are the other triples that do support IFUNCs? Searching a bit more, I find FreeBSD and bionic and have no issue adding these in this PR, are there more?

I wanted to avoid this turning into a long list of libcs but maybe that would be better than the somewhat confusing predicate that I added in #133533. From a quick look:

- Support: glibc, bionic, FreeBSD, DragonflyBSD, NetBSD
- Do not support: musl, OpenBSD, uclibc-ng I guess

So how about we make the logic like this: if the libc supports IFUNC (according to the list above), use IFUNC, otherwise use PAuth relocations. That's still not 100% correct because we will end up using PAuth relocations even when we don't know that the target supports them, but that was the status quo before so maybe that's fine.

> I think that would be wrong, `pauthtest` does not imply any particular libc. Any triple that does not indicate what loader is used should be taken as suggesting a loader _may_ be used that does not support IFUNC.

Okay, I was under the impression that `pauthtest` referred to the musl-based system that was created for testing.

hvdijk commented:
Updated accordingly, does that look okay? I'm a little bit concerned that a global fix like `isOSGlibc()` shouldn't be included in a commit that's meant to be limited to AArch64.

pcc commented on llvm/include/llvm/TargetParser/Triple.h:
```diff
@@ -769,7 +769,7 @@ class Triple {
     return (getOS() == Triple::Linux || getOS() == Triple::KFreeBSD ||
             getOS() == Triple::Hurd) &&
-           !isAndroid();
+           !isAndroid() && !isMusl();
```
Are the other callers ok with this? Looks like it affects ABI stuff on some targets?

kovdan01 reviewed (approved):
LGTM, thanks! This resolves the following linker errors which started to appear after #133533 when building libcxx with signed GOT enabled:

```
ld.lld: error: both AUTH and non-AUTH GOT entries for '_ZTVNSt3__113basic_istreamIcNS_11char_traitsIcEEEE' requested, but only one type of GOT entry per symbol is supported
ld.lld: error: both AUTH and non-AUTH GOT entries for '_ZTVNSt3__113basic_ostreamIcNS_11char_traitsIcEEEE' requested, but only one type of GOT entry per symbol is supported
ld.lld: error: both AUTH and non-AUTH GOT entries for '_ZTVNSt3__114basic_iostreamIcNS_11char_traitsIcEEEE' requested, but only one type of GOT entry per symbol is supported
ld.lld: error: both AUTH and non-AUTH GOT entries for '_ZTVNSt3__113basic_istreamIwNS_11char_traitsIwEEEE' requested, but only one type of GOT entry per symbol is supported
ld.lld: error: both AUTH and non-AUTH GOT entries for '_ZTVNSt3__113basic_ostreamIwNS_11char_traitsIwEEEE' requested, but only one type of GOT entry per symbol is supported
ld.lld: error: both AUTH and non-AUTH GOT entries for '_ZTVNSt3__114basic_ifstreamIcNS_11char_traitsIcEEEE' requested, but only one type of GOT entry per symbol is supported
ld.lld: error: both AUTH and non-AUTH GOT entries for '_ZTVNSt3__114basic_ofstreamIcNS_11char_traitsIcEEEE' requested, but only one type of GOT entry per symbol is supported
```

@pcc I would appreciate if you could take a look at this as soon as you can since your changes from #133533 unfortunately break pauthtest toolchain build with signed GOT enabled, and this PR resolves the issue.

asl commented on llvm/include/llvm/TargetParser/Triple.h:
```diff
@@ -769,7 +769,7 @@ class Triple {
     return (getOS() == Triple::Linux || getOS() == Triple::KFreeBSD ||
             getOS() == Triple::Hurd) &&
-           !isAndroid() && !isMusl();
+           !isAndroid() && !isMusl() && getEnvironment() != Triple::PAuthTest;
```
Please add comment that pauthtest as of now implies not particular libc implementation

asl commented on llvm/test/CodeGen/AArch64/ptrauth-reloc.ll:
```diff
@@ -4,13 +4,13 @@
 ; RUN: llc < ok.ll -mtriple arm64e-apple-darwin \
 ; RUN:   | FileCheck %s --check-prefix=CHECK-MACHO
-; RUN: llc < ok.ll -mtriple aarch64-elf -mattr=+pauth \
+; RUN: llc < ok.ll -mtriple aarch64-android -mattr=+pauth \
```
The tests should remain test elf triple.

pcc commented:
> I took out the removal of `targetSupportsPAuthRelocation` from this PR. It was a good suggestion to look at but in retrospect I don't think it's right, unless I am misunderstanding. For targets that support some PAuth relocations directly, and also support IFUNCs, we should still emit those PAuth relocations directly, that is what that `targetSupportsPAuthRelocation` check achieved before, I don't see how we can do that another way.

I made that suggestion because if a target supports both it doesn't matter what we do.

pcc commented:
I think we should do the simpler thing of only having a condition for IFUNC, especially given that no current target supports both, so the question of what we should do then is theoretical. If a target supports both and needs the slight code size improvement then we can consider adding code to prefer PAuth relocations in that case.

hvdijk commented:
> especially given that no current target supports both

We do have targets where both `targetSupportsPAuthRelocation` and `targetSupportsIRelocativeRelocation` would return `true`. But that just means `targetSupportsPAuthRelocation` was wrong as well, and removing it means we don't need to figure out how to fix it.

pcc commented on llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp:
```diff
@@ -2503,8 +2487,7 @@ const MCExpr *AArch64AsmPrinter::emitPAuthRelocationAsIRelative(
 
   // We only emit an IRELATIVE relocation if the target supports IRELATIVE and
   // does not support the kind of PAuth relocation that we are trying to emit.
```
Update comment

hvdijk committed:
[AArch64] Make IFUNC opt-in rather than opt-out.

IFUNCs require loader support, so for arbitrary environments, the safe
assumption is to assume that they are not supported. In particular,
aarch64-linux-pauthtest may be used with musl, and was wrongly detected
as supporting IFUNCs.

With IFUNC support now being detected more reliably, this also removes
the check for PAuth support. If both are supported, either would work.

pcc reviewed (approved):
> About the `err-disc-elf.ll` -> `err-disc-gnu.ll` test update: that part of the test tests behaviour that's specific to IFUNC relocations, with PAuth relocations it crashes. Which is probably an issue, but also probably not something we need to fix in this specific PR?

Makes sense. Those cases are unrepresentable with PAuth relocations, so if it's crashing we probably want to fix them to error out properly.
